File: org.kevoree.modeling.microframework.typescript/src/main/ts/org/kevoree/modeling/api/json/JSONModelSerializer.ts
///<reference path="../Callback.ts"/>
///<reference path="../KObject.ts"/>
///<reference path="../ModelSerializer.ts"/>
///<reference path="../ModelVisitor.ts"/>
///<reference path="../ThrowableCallback.ts"/>
///<reference path="../VisitResult.ts"/>
///<reference path="../data/AccessMode.ts"/>
///<reference path="../../../../../java/util/Set.ts"/>

class JSONModelSerializer implements ModelSerializer {

  public static KEY_META: string = "@meta";
  public static KEY_UUID: string = "@uuid";
  public static KEY_ROOT: string = "@root";

  public serialize(model: KObject<any,any>, callback: ThrowableCallback<string>): void {
    var builder: StringBuilder = new StringBuilder();
    builder.append("[\n");
    JSONModelSerializer.printJSON(model, builder);
    model.graphVisit(
      public visit(elem: KObject): VisitResult {
        builder.append(",");
        JSONModelSerializer.printJSON(elem, builder);
        return VisitResult.CONTINUE;
      }

, 
      public on(throwable: Throwable): void {
        builder.append("]\n");
        callback.on(builder.toString(), throwable);
      }

);
  }

  public static printJSON(elem: KObject<any,any>, builder: StringBuilder): void {
    builder.append("{\n");
    builder.append("\t\"" + JSONModelSerializer.KEY_META + "\" : \"");
    builder.append(elem.metaClass().metaName());
    builder.append("\",\n");
    builder.append("\t\"" + JSONModelSerializer.KEY_UUID + "\" : \"");
    builder.append(elem.uuid() + "");
    if (elem.isRoot()) {
      builder.append("\",\n");
      builder.append("\t\"" + JSONModelSerializer.KEY_ROOT + "\" : \"");
      builder.append("true");
    }
    builder.append("\",\n");
    for (var i: number = 0; i < elem.metaAttributes().length; i++) {
      var payload: any = elem.get(elem.metaAttributes()[i]);
      if (payload != null) {
        builder.append("\t");
        builder.append("\"");
        builder.append(elem.metaAttributes()[i].metaName());
        builder.append("\" : \"");
        builder.append(payload.toString());
        builder.append("\",\n");
      }
    }
    for (var i: number = 0; i < elem.metaReferences().length; i++) {
      var raw: any[] = elem.view().dimension().universe().storage().raw(elem, AccessMode.READ);
      var payload: any = null;
      if (raw != null) {
        payload = raw[elem.metaReferences()[i].index()];
      }
      if (payload != null) {
        builder.append("\t");
        builder.append("\"");
        builder.append(elem.metaReferences()[i].metaName());
        builder.append("\" :");
        if (elem.metaReferences()[i].single()) {
          builder.append("\"");
          builder.append(payload.toString());
          builder.append("\"");
        } else {
          var elems: Set<number> = <Set<number>>payload;
          var elemsArr: number[] = elems.toArray(new Array());
          var isFirst: boolean = true;
          builder.append(" [");
          for (var j: number = 0; j < elemsArr.length; j++) {
            if (!isFirst) {
              builder.append(",");
            }
            builder.append("\"");
            builder.append(elemsArr[j] + "");
            builder.append("\"");
            isFirst = false;
          }
          builder.append("]");
        }
        builder.append(",\n");
      }
    }
    builder.append("}\n");
  }

}

File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/format/json/JsonModelSerializer.java
package org.kevoree.modeling.format.json;

import org.kevoree.modeling.KModelVisitor;
import org.kevoree.modeling.KObject;
import org.kevoree.modeling.Callback;
import org.kevoree.modeling.KVisitResult;
import org.kevoree.modeling.abs.AbstractKObject;
import org.kevoree.modeling.memory.KCacheElementSegment;
import org.kevoree.modeling.memory.struct.segment.HeapCacheSegment;
import org.kevoree.modeling.memory.AccessMode;
import org.kevoree.modeling.memory.manager.JsonRaw;

public class JsonModelSerializer {

    public static void serialize(final KObject model, final Callback<String> callback) {
        ((AbstractKObject) model)._manager.getRoot(model.universe(), model.now(), new Callback<KObject>() {
            @Override
            public void on(final KObject rootObj) {
                boolean isRoot = false;
                if (rootObj != null) {
                    isRoot = rootObj.uuid() == model.uuid();
                }
                final StringBuilder builder = new StringBuilder();
                builder.append("[\n");
                printJSON(model, builder, isRoot);
                model.visit(new KModelVisitor() {
                    @Override
                    public KVisitResult visit(KObject elem) {
                        boolean isRoot2 = false;
                        if (rootObj != null) {
                            isRoot2 = rootObj.uuid() == elem.uuid();
                        }
                        builder.append(",\n");
                        try {
                            printJSON(elem, builder, isRoot2);
                        } catch (Exception e) {
                            e.printStackTrace();
                            builder.append("{}");
                        }
                        return KVisitResult.CONTINUE;
                    }
                },new Callback<Throwable>() {
                    @Override
                    public void on(Throwable throwable) {
                        builder.append("\n]\n");
                        callback.on(builder.toString());
                    }
                });
            }
        });
    }

    public static void printJSON(KObject elem, StringBuilder builder, boolean isRoot) {
        if (elem != null) {
            KCacheElementSegment raw = ((AbstractKObject) elem)._manager.segment(elem, AccessMode.READ);
            if (raw != null) {
                builder.append(JsonRaw.encode(raw, elem.uuid(), elem.metaClass(), isRoot));
            }
        }
    }

}

